src/Microsoft.Data.Analysis.Interactive/DataFrameKernelExtension.cs

```diff
@@ -1,47 +1,224 @@
+            prettyFormatter.AppendLine();
+            prettyFormatter.Append(buildNamedDataFrame);
+            await context.DisplayAsync(prettyFormatter.ToString());
+            context.Publish(new DisplayedValueProduced($"Created {typeName} {dataFrameName}: ", context.Command));
+
+            await SubmitCodeToKernel(strBuilder, context);
+        }
+
+        public StringBuilder GenerateTypedDataFrame(DataFrame df, string typeName, KernelInvocationContext context, out StringBuilder prettyFormatter)
+        {
+            StringBuilder typedDataFrame = GetTypedDataFrameWithProperties(df, typeName, out prettyFormatter);
+            return typedDataFrame;
+        }
+
+        private async Task SubmitCodeToKernel(StringBuilder code, KernelInvocationContext context)
+        {
+            var command = new SubmitCode(code.ToString());
+            await context.HandlingKernel.SendAsync(command);
+        }
+
+        public async Task HandleDataFrameAsync(string dataFrameName, string typeName, KernelInvocationContext context)
+        {
+            if (context.HandlingKernel is CSharpKernel cSharp)
+            {
+                System.Collections.Immutable.ImmutableArray<CodeAnalysis.Scripting.ScriptVariable> variables = cSharp.ScriptState.Variables;
+                for (int i = 0; i < variables.Length; i++)
+                {
+                    CodeAnalysis.Scripting.ScriptVariable variable = variables[i];
+                    if ((dataFrameName == null || variable.Name == dataFrameName) && variable.Value is DataFrame df)
+                    {
+                        var strBuilder = GenerateTypedDataFrame(df, typeName, context, out StringBuilder prettyFormatter);
+                        await context.DisplayAsync(prettyFormatter.ToString());
+                        await SubmitCodeToKernel(strBuilder, context);
+                    }
+                }
+            }
+        }
+
+        public Task OnLoadAsync(IKernel kernel)
+        {
             var kernelBase = kernel as KernelBase;
-            var directive = new Command("#!doit")
+            var directive = new Command("#!generatedataframe")
             {
-                Handler = CommandHandler.Create(async (FileInfo csv, string typeName, KernelInvocationContext context) =>
+                Handler = CommandHandler.Create(async (FileInfo csv, string dataFrameName, string typeName, KernelInvocationContext context) =>
                 {
                     // do the job
-                    var command = new SubmitCode(@$"public class {typeName}{{}}");
-                    context.Publish(new DisplayedValueProduced($"emitting {typeName} from {csv.FullName}", context.Command));
-                    await context.HandlingKernel.SendAsync(command);
+                    try
+                    {
+                        if (csv != null)
+                        {
+                            HandleCsvAsync(csv, context);
```

Comment:
Why doesn't `HandleCsvAsync` get the `--type-name` and `--dataframe-name` parameters passed to it?